devserver_lib/src/lib.rs

```diff
@@ -44,46 +63,28 @@ fn handle_client<T: Read + Write>(mut stream: T, root_path: &str, reload: bool,
     let path = parts.next().unwrap().trim();
     let _http_version = parts.next().unwrap().trim();
 
-    // Replace white space characters with proper whitespace and remove any paths that refer to the parent.
-    let path = path.replace("../", "").replace("%20", " ");
-    let path = if path.ends_with("/") {
-        Path::new(root_path).join(Path::new(&format!(
-            "{}{}",
-            path.trim_start_matches('/'),
-            "index.html"
-        )))
-    } else {
-        Path::new(root_path).join(path.trim_matches('/'))
-    };
-
-    let extension = path.extension().and_then(OsStr::to_str);
-
-    let (file_contents, extension) = if extension != None {
-        (fs::read(&path), extension)
-    } else {
-        // If the request has no extension look first for a matching file without an extension
-        if let Ok(file_contents) = fs::read(&path) {
-            println!("WARNING: Serving file without extension: [ {} ] with media type 'application/octet-stream'", &path.to_str().unwrap());
-            (Ok(file_contents), None)
+    let file_contents: Result<Vec<u8>>;
+    let candidates = file_candidates(path);
+    let blank = FileProxy::blank();
+    let chosen = candidates.iter().find(|candidate| {
+        if let Ok(file_contents) = candidate.read(root_path) {
```

Comment:
Based on your usage I think you have a small misconception about how this line is working, which makes sense because it's weird syntax.

```rust
if let Ok(file_contents) = candidate.read(root_path) {
```
is declaring a new variable called `file_contents` whose scope limited to this if statement. It's **not** assigning to the existing `file_contents`. It's shadowing the existing variable.

So what you want is more like this:
```rust
if let Ok(new_file_contents) = candidate.read(root_path) {
   file_contents = new_file_contents;
   true
}
...
```